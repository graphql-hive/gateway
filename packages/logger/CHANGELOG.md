# @graphql-hive/logger

## 1.0.4
### Patch Changes



- [#1523](https://github.com/graphql-hive/gateway/pull/1523) [`bb2621c`](https://github.com/graphql-hive/gateway/commit/bb2621ce85c42ccbc97c6ca128f959bcb2bb6475) Thanks [@dependabot](https://github.com/apps/dependabot)! - dependencies updates:
  
  - Updated dependency [`@logtape/logtape@^1.1.1` ↗︎](https://www.npmjs.com/package/@logtape/logtape/v/1.1.1) (from `^1.1.0`, in `peerDependencies`)
  - Updated dependency [`pino@^9.10.0` ↗︎](https://www.npmjs.com/package/pino/v/9.10.0) (from `^9.9.5`, in `peerDependencies`)

## 1.0.3
### Patch Changes



- [#1484](https://github.com/graphql-hive/gateway/pull/1484) [`950ebd1`](https://github.com/graphql-hive/gateway/commit/950ebd1d1686846b59b555695c1738e25fd3268e) Thanks [@dependabot](https://github.com/apps/dependabot)! - dependencies updates:
  
  - Updated dependency [`@logtape/logtape@^1.0.5` ↗︎](https://www.npmjs.com/package/@logtape/logtape/v/1.0.5) (from `^1.0.4`, in `peerDependencies`)
  - Updated dependency [`pino@^9.9.5` ↗︎](https://www.npmjs.com/package/pino/v/9.9.5) (from `^9.9.4`, in `peerDependencies`)


- [#1495](https://github.com/graphql-hive/gateway/pull/1495) [`fe99f74`](https://github.com/graphql-hive/gateway/commit/fe99f74dd11fdf2928ca7080d4d2e5dfd1e2f18e) Thanks [@dependabot](https://github.com/apps/dependabot)! - dependencies updates:
  
  - Updated dependency [`@logtape/logtape@^1.1.0` ↗︎](https://www.npmjs.com/package/@logtape/logtape/v/1.1.0) (from `^1.0.5`, in `peerDependencies`)

## 1.0.2
### Patch Changes



- [#1452](https://github.com/graphql-hive/gateway/pull/1452) [`b0e5568`](https://github.com/graphql-hive/gateway/commit/b0e55688d4fc22d0bfbf664de52e78e9642d7014) Thanks [@dependabot](https://github.com/apps/dependabot)! - dependencies updates:
  
  - Updated dependency [`@logtape/logtape@^1.0.4` ↗︎](https://www.npmjs.com/package/@logtape/logtape/v/1.0.4) (from `^1.0.0`, in `peerDependencies`)
  - Updated dependency [`pino@^9.9.4` ↗︎](https://www.npmjs.com/package/pino/v/9.9.4) (from `^9.6.0`, in `peerDependencies`)


- [#1452](https://github.com/graphql-hive/gateway/pull/1452) [`b0e5568`](https://github.com/graphql-hive/gateway/commit/b0e55688d4fc22d0bfbf664de52e78e9642d7014) Thanks [@dependabot](https://github.com/apps/dependabot)! - Relax `PinoLogWriter`'s typings for `FastifyLogger` compatibility

## 1.0.1
### Patch Changes



- [#956](https://github.com/graphql-hive/gateway/pull/956) [`46d2661`](https://github.com/graphql-hive/gateway/commit/46d26615c2c3c5f936c1d1bca1d03b025c1ce86a) Thanks [@EmrysMyrddin](https://github.com/EmrysMyrddin)! - Introducing Hive Logger
  
  [Read more about it on the Hive Logger documentation website.](https://the-guild.dev/graphql/hive/docs/logger)
