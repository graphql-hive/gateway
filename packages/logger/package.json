{
  "name": "@graphql-hive/logger",
  "version": "0.0.0",
  "type": "module",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/graphql-hive/gateway.git",
    "directory": "packages/logger"
  },
  "author": {
    "email": "contact@the-guild.dev",
    "name": "The Guild",
    "url": "https://the-guild.dev"
  },
  "license": "MIT",
  "engines": {
    "node": ">=18.0.0"
  },
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      },
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.js"
      }
    },
    "./request": {
      "require": {
        "types": "./dist/request.d.cts",
        "default": "./dist/request.cjs"
      },
      "import": {
        "types": "./dist/request.d.ts",
        "default": "./dist/request.js"
      }
    },
    "./writers/pino": {
      "require": {
        "types": "./dist/writers/pino.d.cts",
        "default": "./dist/writers/pino.cjs"
      },
      "import": {
        "types": "./dist/writers/pino.d.ts",
        "default": "./dist/writers/pino.js"
      }
    },
    "./package.json": "./package.json"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "pkgroll --clean-dist",
    "prepack": "yarn build"
  },
  "peerDependencies": {
    "pino": "^9.6.0"
  },
  "peerDependenciesMeta": {
    "pino": {
      "optional": true
    }
  },
  "devDependencies": {
    "@types/quick-format-unescaped": "^4.0.3",
    "@whatwg-node/disposablestack": "^0.0.6",
    "fast-safe-stringify": "^2.1.1",
    "pino": "^9.6.0",
    "pkgroll": "2.11.2",
    "quick-format-unescaped": "^4.0.4"
  },
  "sideEffects": false,
  "dependencies.info": "all of the dependencies are in devDependencies which will bundle them into the package using pkgroll making pkgroll ultimately zero-dep with a smaller footprint because of tree-shaking"
}
