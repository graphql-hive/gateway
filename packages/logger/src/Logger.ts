import format from 'quick-format-unescaped';
import {
  Attributes,
  getEnv,
  isPromise,
  logLevel,
  MaybeLazy,
  parseAttrs,
  shouldLog,
  truthyEnv,
} from './utils';
import { ConsoleLogWriter, LogWriter } from './writers';

export type LogLevel = 'trace' | 'debug' | 'info' | 'warn' | 'error';

// TODO: explain what happens when attribute keys match existing keys from the logger (like "msg")

// TODO: an "id" or "name" of a logger allowing us to create scoped loggers which on their own can be disabled/enabled

export interface LoggerOptions {
  /**
   * The minimum log level to log.
   *
   * Providing `false` will disable all logging.
   *
   * @default env.LOG_LEVEL || env.DEBUG ? 'debug' : 'info'
   */
  level?: LogLevel | false;
  /** A prefix to include in every log's message. */
  prefix?: string;
  /**
   * The attributes to include in all logs. Is mainly used to pass the parent
   * attributes when creating {@link Logger.child child loggers}.
   */
  attrs?: MaybeLazy<Attributes>;
  /**
   * The log writers to use when writing logs.
   *
   * @default [new ConsoleLogWriter()]
   */
  writers?: [LogWriter, ...LogWriter[]];
}

export class Logger implements LogWriter {
  #level: LogLevel | false;
  #prefix: string | undefined;
  #attrs: MaybeLazy<Attributes> | undefined;
  #writers: [LogWriter, ...LogWriter[]];
  #pendingWrites = new Set<Promise<void>>();

  constructor(opts: LoggerOptions = {}) {
    let logLevelEnv = getEnv('LOG_LEVEL');
    if (logLevelEnv && !(logLevelEnv in logLevel)) {
      throw new Error(
        `Invalid LOG_LEVEL environment variable "${logLevelEnv}". Must be one of: ${[...Object.keys(logLevel), 'false'].join(',  ')}`,
      );
    }
    this.#level =
      opts.level ??
      (logLevelEnv as LogLevel) ??
      (truthyEnv('DEBUG') ? 'debug' : 'info');
    this.#prefix = opts.prefix;
    this.#attrs = opts.attrs;
    this.#writers = opts.writers ?? [new ConsoleLogWriter()];
  }

  public get prefix() {
    return this.#prefix;
  }

  public get level() {
    return this.#level;
  }

  public write(
    level: LogLevel,
    attrs: Attributes | null | undefined,
    msg: string | null | undefined,
  ): void {
    const pendingWrites = this.#writers
      .map((writer) => writer.write(level, attrs, msg))
      .filter(isPromise);

    for (const pendingWrite of pendingWrites) {
      this.#pendingWrites.add(pendingWrite);
      pendingWrite.catch(() => {
        // TODO: what to do if the async write failed?
      });
      pendingWrite.finally(() => this.#pendingWrites.delete(pendingWrite));
    }
  }

  public flush() {
    if (this.#pendingWrites.size) {
      return Promise.all(this.#pendingWrites).then(() => {
        // void
      });
    }
    return;
  }

  // TODO: flush on dispose

  //

  public child(prefix: string): Logger;
  public child(attrs: MaybeLazy<Attributes>, prefix?: string): Logger;
  public child(
    prefixOrAttrs: string | MaybeLazy<Attributes>,
    prefix?: string,
  ): Logger {
    if (typeof prefixOrAttrs === 'string') {
      return new Logger({
        prefix: prefixOrAttrs,
        writers: this.#writers,
      });
    }
    return new Logger({
      prefix,
      attrs: prefixOrAttrs,
      writers: this.#writers,
    });
  }

  //

  public log(level: LogLevel): void;
  public log(level: LogLevel, attrs: MaybeLazy<Attributes>): void;
  public log(level: LogLevel, msg: string, ...interpol: unknown[]): void;
  public log(
    level: LogLevel,
    attrs: MaybeLazy<Attributes>,
    msg: string,
    ...interpol: unknown[]
  ): void;
  public log(
    level: LogLevel,
    maybeAttrsOrMsg?: MaybeLazy<Attributes> | string | null | undefined,
    ...rest: unknown[]
  ): void {
    if (!shouldLog(this.#level, level)) {
      return;
    }

    let msg: string | undefined;
    let attrs: MaybeLazy<Attributes> | undefined;
    if (typeof maybeAttrsOrMsg === 'string') {
      msg = maybeAttrsOrMsg;
    } else if (maybeAttrsOrMsg) {
      attrs = maybeAttrsOrMsg;
      if (typeof rest[0] === 'string') {
        // we shift because the "rest" becomes "interpol"
        msg = rest.shift() as string;
      }
    }

    if (this.#prefix) {
      msg = `${this.#prefix.trim()} ${msg || ''}`.trim(); // we trim everything because maybe the "msg" is empty
    }

    attrs = attrs ? parseAttrs(attrs) : attrs;
    attrs = this.#attrs ? { ...parseAttrs(this.#attrs), ...attrs } : attrs;
    msg = msg ? format(msg, rest) : msg;

    this.write(level, attrs, msg);
    if (truthyEnv('LOG_TRACE_LOGS')) {
      console.trace('ðŸ‘†');
    }
  }

  public trace(): void;
  public trace(attrs: MaybeLazy<Attributes>): void;
  public trace(msg: string, ...interpol: unknown[]): void;
  public trace(
    attrs: MaybeLazy<Attributes>,
    msg: string,
    ...interpol: unknown[]
  ): void;
  public trace(...args: any): void {
    this.log(
      'trace',
      // @ts-expect-error
      ...args,
    );
  }

  public debug(): void;
  public debug(attrs: MaybeLazy<Attributes>): void;
  public debug(msg: string, ...interpol: unknown[]): void;
  public debug(
    attrs: MaybeLazy<Attributes>,
    msg: string,
    ...interpol: unknown[]
  ): void;
  public debug(...args: any): void {
    this.log(
      'debug',
      // @ts-expect-error
      ...args,
    );
  }

  public info(): void;
  public info(attrs: MaybeLazy<Attributes>): void;
  public info(msg: string, ...interpol: unknown[]): void;
  public info(
    attrs: MaybeLazy<Attributes>,
    msg: string,
    ...interpol: unknown[]
  ): void;
  public info(...args: any): void {
    this.log(
      'info',
      // @ts-expect-error
      ...args,
    );
  }

  public warn(): void;
  public warn(attrs: MaybeLazy<Attributes>): void;
  public warn(msg: string, ...interpol: unknown[]): void;
  public warn(
    attrs: MaybeLazy<Attributes>,
    msg: string,
    ...interpol: unknown[]
  ): void;
  public warn(...args: any): void {
    this.log(
      'warn',
      // @ts-expect-error
      ...args,
    );
  }

  public error(): void;
  public error(attrs: MaybeLazy<Attributes>): void;
  public error(msg: string, ...interpol: unknown[]): void;
  public error(
    attrs: MaybeLazy<Attributes>,
    msg: string,
    ...interpol: unknown[]
  ): void;
  public error(...args: any): void {
    this.log(
      'error',
      // @ts-expect-error
      ...args,
    );
  }
}
