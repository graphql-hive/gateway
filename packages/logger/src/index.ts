export * from './logger';
export * from './writers';
/** @deprecated Please migrate to using the './Logger' instead. */
export * from './legacyLogger';
