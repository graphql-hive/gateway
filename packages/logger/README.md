# Hive Logger

Lightweight and customizable logging utility designed for use within the GraphQL Hive ecosystem. It provides structured logging capabilities, making it easier to debug and monitor applications effectively.

## Compatibility

The Hive Logger is designed to work seamlessly in all JavaScript environments, including Node.js, browsers, and serverless platforms. Its lightweight design ensures minimal overhead, making it suitable for a wide range of applications.

# Getting Started

## Install

```sh
npm i @graphql-hive/logger
```

## Basic Usage

Create a default logger that set to the `info` log level writing to the console.

```ts
import { Logger } from '@graphql-hive/logger';

const log = new Logger();

log.debug('I wont be logged by default');

log.info({ some: 'attributes' }, 'Hello %s!', 'world');

const child = log.child({ requestId: '123-456' });

child.warn({ more: 'attributes' }, 'Oh hello child!');

const err = new Error('Woah!');

child.error({ err }, 'Something went wrong!');
```

Will produce the following output to the console output:

<!-- prettier-ignore-start -->
```sh
2025-04-10T14:00:00.000Z INF Hello world!
  some: "attributes"
2025-04-10T14:00:00.000Z WRN Oh hello child!
  requestId: "123-456"
  more: "attributes"
2025-04-10T14:00:00.000Z ERR Something went wrong!
  requestId: "123-456"
  err: {
    stack: "Error: Woah!
        at <anonymous> (/project/example.js:13:13)
        at ModuleJob.run (node:internal/modules/esm/module_job:274:25)
        at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:644:26)
        at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:98:5)"
    message: "Woah!"
    name: "Error"
    class: "Error"
  }
```
<!-- prettier-ignore-end -->

or if you wish to have JSON output, set the `LOG_JSON` environment variable to a truthy value:

<!-- prettier-ignore-start -->
```sh
$ LOG_JSON=1 node example.js

{"some":"attributes","level":"info","msg":"Hello world!","timestamp":"2025-04-10T14:00:00.000Z"}
{"requestId":"123-456","more":"attributes","level":"info","msg":"Hello child!","timestamp":"2025-04-10T14:00:00.000Z"}
{"requestId":"123-456","err":{"stack":"Error: Woah!\n    at <anonymous> (/project/example.js:13:13)\n    at ModuleJob.run (node:internal/modules/esm/module_job:274:25)\n    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:644:26)\n    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:98:5)","message":"Woah!","name":"Error","class":"Error"},"level":"error","msg":"Something went wrong!","timestamp":"2025-04-10T14:00:00.000Z"}
```
<!-- prettier-ignore-end -->

## Message Formatting

The Hive Logger uses the [`quick-format-unescaped` library](https://github.com/pinojs/quick-format-unescaped) to format log messages that include interpolation (e.g., placeholders like %s, %d, etc.).

```ts
import { Logger } from '@graphql-hive/logger';

const log = new Logger();

log.info('hello %s %j %d %o', 'world', { obj: true }, 4, { another: 'obj' });
```

Outputs:

```sh
2025-04-10T14:00:00.000Z INF hello world {"obj":true} 4 {"another":"obj"}
```

## Logging Levels

The default logger uses the `info` log level which will make sure to log only `info`+ logs. Available log levels are:

- `trace`
- `debug`
- `info` _default_
- `warn`
- `error`

You can change the loggers logging level on creation or dynamically.

```ts
import { Logger } from '@graphql-hive/logger';

const log = new Logger({ level: 'debug' });

log.trace(
  // you can suply "lazy" attributes which wont be evaluated unless the log level allows logging
  () => ({
    wont: 'be evaluated',
    some: expensiveOperation(),
  }),
  'Wont be logged and attributes wont be evaluated',
);

log.debug('Hello world!');

const child = log.child('[prefix] ');

child.debug('Child loggers inherit the parent log level');

log.setLevel('trace');

log.trace(() => ({ hi: 'there' }), 'Now tracing is logged too!');

child.trace('Also on the child logger');

child.setLevel('info');

log.trace('Still logging!');

child.debug('Wont be logged because the child has a different log level now');

child.info('Hello child!');
```

Outputs the following to the console:

<!-- prettier-ignore-start -->
```sh
2025-04-10T14:00:00.000Z DBG Hello world!
2025-04-10T14:00:00.000Z DBG [prefix] Child loggers inherit the parent log level
2025-04-10T14:00:00.000Z TRC Now tracing is logged too!
  hi: "there"
2025-04-10T14:00:00.000Z TRC [prefix] Also on the child logger
2025-04-10T14:00:00.000Z TRC Still logging!
2025-04-10T14:00:00.000Z INF Hello child!
```
<!-- prettier-ignore-end -->

## Writers

Logger writers are responsible for handling how and where log messages are output. In Hive Logger, writers are pluggable components that receive structured log data and determine its final destination and format. This allows you to easily customize logging behavior, such as printing logs to the console, writing them as JSON, storing them in memory for testing, or sending them to external systems.

By default, Hive Logger provides several built-in writers, but you can also implement your own to suit your application's needs. The built-ins are:

### `MemoryLogWriter`

Writes the logs to memory allowing you to access the logs. Mostly useful for testing.

```ts
import { Logger, MemoryLogWriter } from '@graphql-hive/logger';

const writer = new MemoryLogWriter();

const log = new Logger({ writers: [writer] });

log.info({ my: 'attrs' }, 'Hello World!');

console.log(writer.logs);
```

Outputs:

```sh
[ { level: 'info', msg: 'Hello World!', attrs: { my: 'attrs' } } ]
```

### `ConsoleLogWriter` (default)

The default log writer used by the Hive Logger. It outputs log messages to the console in a human-friendly, colorized format, making it easy to distinguish log levels and read structured attributes. Each log entry includes a timestamp, the log level (with color), the message, and any additional attributes (with colored keys), which are pretty-printed and formatted for clarity.

The writer works in both Node.js and browser-like environments, automatically disabling colors if not supported. This makes `ConsoleLogWriter` ideal for all cases, providing clear and readable logs out of the box.

```ts
import { ConsoleLogWriter, Logger } from '@graphql-hive/logger';

const writer = new ConsoleLogWriter({
  noColor: true, // defaults to env.NO_COLOR. read more: https://no-color.org/
  noTimestamp: true,
});

const log = new Logger({ writers: [writer] });

log.info({ my: 'attrs' }, 'Hello World!');
```

Outputs:

<!-- prettier-ignore-start -->
```sh
INF Hello World!
  my: "attrs"
```
<!-- prettier-ignore-end -->

### `JSONLogWriter` (default when `LOG_JSON=1`)

Built-in log writer that outputs each log entry as a structured JSON object. When used, it prints logs to the console in JSON format, including all provided attributes, the log level, message, and a timestamp.

If the `LOG_JSON_PRETTY=1` environment variable is provided, the output will be pretty-printed for readability; otherwise, it is compact.

This writer's format is ideal for machine parsing, log aggregation, or integrating with external logging systems, especially useful for production environments or when logs need to be consumed by other tools.

```ts
import { JSONLogWriter, Logger } from '@graphql-hive/logger';

const log = new Logger({ writers: [new JSONLogWriter()] });

log.info({ my: 'attrs' }, 'Hello World!');
```

Outputs:

<!-- prettier-ignore-start -->
```sh
{"my":"attrs","level":"info","msg":"Hello World!","timestamp":"2025-04-10T14:00:00.000Z"}
```
<!-- prettier-ignore-end -->

Or pretty printed:

<!-- prettier-ignore-start -->
```sh
$ LOG_JSON_PRETTY=1 node example.js

{
  "my": "attrs",
  "level": "info",
  "msg": "Hello World!",
  "timestamp": "2025-04-10T14:00:00.000Z"
}
```
<!-- prettier-ignore-end -->

### Custom Writers

You can implement custom log writers for the Hive Logger by creating a class that implements the `LogWriter` interface. This interface requires a single `write` method, which receives the log level, attributes, and message.

Your writer can perform any action, such as sending logs to a file, external service, or custom destination.

Writers can be synchronous (returning `void`) or asynchronous (returning a `Promise<void>`). If your writer performs asynchronous operations (like network requests or file writes), simply return a promise from the `write` method.

```ts
import {
  Attributes,
  ConsoleLogWriter,
  Logger,
  LogLevel,
  LogWriter,
} from '@graphql-hive/logger';

class HTTPLogWriter implements LogWriter {
  async write(level: LogLevel, attrs: Attributes, msg: string) {
    await fetch('https://my-log-service.com', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ level, attrs, msg }),
    });
  }
}

const log = new Logger({
  // send logs both to the HTTP loggging service and output them to the console
  writers: [new HTTPLogWriter(), new ConsoleLogWriter()],
});

log.info('Hello World!');

await log.flush(); // make sure all async writes settle
```

#### Flushing and Non-Blocking Logging

The logger does not block when you log asynchronously. Instead, it tracks all pending async writes internally. When you call `log.flush()` or dispose the logger when using the [Explicit Resource Management](https://github.com/tc39/proposal-explicit-resource-management), it waits for all pending writes to finish, ensuring no logs are lost on shutdown. During normal operation, logging remains fast and non-blocking, even if some writers are async.

This design allows you to use async writers without impacting the performance of your application or blocking the main thread.

##### Handling Async Write Errors

The Logger handles write errors for asynchronous writers by tracking all write promises. When `await log.flush()` is called (including during async disposal), it waits for all pending writes to settle. If any writes fail (i.e., their promises reject), their errors are collected and after all writes have settled, if there were any errors, an `AggregateError` is thrown containing all the individual write errors.

```ts
import { Logger } from './Logger';

let i = 0;
const log = new Logger({
  writers: [
    {
      async write() {
        i++;
        throw new Error('Write failed! #' + i);
      },
    },
  ],
});

// no fail during logs
log.info('hello');
log.info('world');

try {
  await log.flush();
} catch (e) {
  // flush will fail with each individually failed writes
  console.error(e);
}
```

Outputs:

```sh
AggregateError: Failed to flush 2 writes
    at async <anonymous> (/project/example.js:20:3) {
  [errors]: [
    Error: Write failed! #1
        at Object.write (/project/example.js:9:15),
    Error: Write failed! #2
        at Object.write (/project/example.js:9:15)
  ]
}
```
