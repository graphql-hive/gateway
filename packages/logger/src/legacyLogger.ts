import { Logger, LogLevel } from './logger';
import { shouldLog } from './utils';

// type comes from "@graphql-mesh/types" package, we're copying them over just to avoid including the whole package
export type LazyLoggerMessage = (() => any | any[]) | any;

/** @deprecated Please migrate to using the {@link Logger} instead.*/
export class LegacyLogger {
  #logger: Logger;

  constructor(logger: Logger) {
    this.#logger = logger;
  }

  static from(logger: Logger): LegacyLogger {
    return new LegacyLogger(logger);
  }

  #log(level: LogLevel, ...[maybeMsgOrArg, ...restArgs]: any[]) {
    if (typeof maybeMsgOrArg === 'string') {
      if (restArgs.length) {
        this.#logger.log(level, restArgs, maybeMsgOrArg);
      } else {
        this.#logger.log(level, maybeMsgOrArg);
      }
    } else {
      if (restArgs.length) {
        this.#logger.log(level, [maybeMsgOrArg, ...restArgs]);
      } else {
        this.#logger.log(level, maybeMsgOrArg);
      }
    }
  }

  log(...args: any[]) {
    this.#log('info', ...args);
  }

  warn(...args: any[]) {
    this.#log('warn', ...args);
  }

  info(...args: any[]) {
    this.#log('info', ...args);
  }

  error(...args: any[]) {
    this.#log('error', ...args);
  }

  debug(...lazyArgs: LazyLoggerMessage[]) {
    if (!shouldLog(this.#logger.level, 'debug')) {
      // we only return early here because only debug can have lazy logs
      return;
    }
    this.#log('debug', ...handleLazyMessage(lazyArgs));
  }

  child(name: string | Record<string, string | number>): LegacyLogger {
    name =
      stringifyName(name) +
      // append space if object is strigified to space out the prefix
      (typeof name === 'object' ? ' ' : '');
    if (this.#logger.prefix === name) {
      return this;
    }
    return LegacyLogger.from(this.#logger.child(name));
  }

  addPrefix(prefix: string | Record<string, string | number>): LegacyLogger {
    prefix = stringifyName(prefix);
    if (this.#logger.prefix?.includes(prefix)) {
      // TODO: why do we do this?
      return this;
    }
    return LegacyLogger.from(this.#logger.child(prefix));
  }
}

function stringifyName(name: string | Record<string, string | number>) {
  if (typeof name === 'string' || typeof name === 'number') {
    return `${name}`;
  }
  const names: string[] = [];
  for (const [key, value] of Object.entries(name)) {
    names.push(`${key}=${value}`);
  }
  return `${names.join(', ')}`;
}

function handleLazyMessage(lazyArgs: LazyLoggerMessage[]) {
  return lazyArgs.flat(Infinity).flatMap((arg) => {
    if (typeof arg === 'function') {
      return arg();
    }
    return arg;
  });
}
