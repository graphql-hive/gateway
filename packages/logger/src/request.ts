import { Logger } from './logger';

export const requestIdByRequest = new WeakMap<Request, string>();

const loggerByRequest = new WeakMap<Request, Logger>();

/**
 * Gets the {@link Logger} of for the {@link request}.
 *
 * If the request does not have a logger, the provided {@link log}
 * will be associated to the {@link request} and returned.
 */
export function loggerForRequest(log: Logger, request: Request): Logger {
  const reqLog = loggerByRequest.get(request);
  if (reqLog) {
    return reqLog;
  }
  loggerByRequest.set(request, log);
  return log;
}
