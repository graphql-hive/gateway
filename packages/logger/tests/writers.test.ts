import { describe, expect, it } from 'vitest';
import { Logger } from '../src/logger';
import {
  ConsoleLogWriter,
  ConsoleLogWriterOptions,
  jsonStringify,
} from '../src/writers';

describe('ConsoleLogWriter', () => {
  function createTConsoleLogger(opts?: Partial<ConsoleLogWriterOptions>) {
    const logs: string[] = [];
    const writer = new ConsoleLogWriter({
      console: {
        debug: (...args: unknown[]) => {
          logs.push(args.map((arg) => jsonStringify(arg)).join(' '));
        },
        info: (...args: unknown[]) => {
          logs.push(args.map((arg) => jsonStringify(arg)).join(' '));
        },
        warn: (...args: unknown[]) => {
          logs.push(args.map((arg) => jsonStringify(arg)).join(' '));
        },
        error: (...args: unknown[]) => {
          logs.push(args.map((arg) => jsonStringify(arg)).join(' '));
        },
      },
      noTimestamp: true,
      noColor: true,
      ...opts,
    });
    return [new Logger({ level: 'trace', writers: [writer] }), logs] as const;
  }

  it('should pretty print the attributes', () => {
    const [log, logs] = createTConsoleLogger();

    log.trace({ a: 1, b: 2 }, 'obj');
    log.debug(['a', 'b', 'c'], 'arr');
    log.info({ a: { b: { c: { d: 1 } } } }, 'nested');
    log.warn([{ a: 1 }, { b: 2 }], 'arr objs');
    log.error({ str: 'a\nb\nc', err: { message: 'woah!' } }, 'multlinestring');

    log.info(
      {
        query: `
{
  hi(howMany: 1) {
    hello
    world
  }
}
`,
      },
      'graphql',
    );

    expect(logs).toMatchSnapshot();
  });

  it('should color levels and keys', () => {
    const [log, logs] = createTConsoleLogger({ noColor: false });

    log.trace({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');
    log.debug({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');
    log.info({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');
    log.warn({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');
    log.error({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');

    expect(logs).toMatchSnapshot();
  });

  it('should flush async logs', async () => {
    const [log, logs] = createTConsoleLogger({ noColor: true, async: true });

    log.trace({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');
    log.debug({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');
    log.info({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');
    log.warn({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');
    log.error({ hello: { dear: 'world', try: ['num', 1, 2] } }, 'hi');

    expect(logs).toHaveLength(0);
    await log.flush();
    expect(logs).toMatchSnapshot();
  });
});
