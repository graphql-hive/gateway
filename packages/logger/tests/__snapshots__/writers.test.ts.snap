// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsoleLogWriter should color levels and keys 1`] = `
[
  "[36mTRC[0m [1mhi[0m 
  [35mhello:[0m "world"",
  "[90mDBG[0m [1mhi[0m 
  [35mhello:[0m "world"",
  "[32mINF[0m [1mhi[0m 
  [35mhello:[0m "world"",
  "[33mWRN[0m [1mhi[0m 
  [35mhello:[0m "world"",
  "[41;39mERR[0m [1mhi[0m 
  [35mhello:[0m "world"",
]
`;

exports[`ConsoleLogWriter should pretty print the attributes 1`] = `
[
  "TRC obj 
  a: 1
  b: 2",
  "DBG arr 
  "a"
  "b"
  "c"",
  "INF nested 
  a: {
    b: {
      c: {
        d: 1
      }
    }
  }",
  "WRN arr objs 
  {
    a: 1
  }
  {
    b: 2
  }",
  "ERR multlinestring 
  str: "a
  b
  c"
  err: {
    stack: "<stack>"
    message: "woah!"
    name: "Error"
    class: "Error"
  }",
  "INF graphql 
  query: "
  {
    hi(howMany: 1) {
      hello
      world
    }
  }
  "",
]
`;
